### DISCUSSION TITLE
Additional Feature Collection

### DISCUSSION THREAD

**Discussion by noctarius** [sombriks/node-libgpiod#9] [Ideas]
@dmitrydvorkin, @sombriks let's move the discussion over here and start a collection of ideas :-)

- Remove dependency on libgpiod and use ioctl directly
- With the above comes the requirement to implement the gpio character device api V1 and V2
- Functionality to return all available gpio chips
- Functionality to return the pinmux configuration (bias, and stuff) and pinmux name
- Functionality to return the gpio number, e.g. on IMX6 and some others: (chipNumber - 1) * 32 + lineOffset
- Functionality to configure Edge detection (with event triggering)
- Bulk operations

**Comment by noctarius** [comment]
Tbh, I don't really care for the removal of the libgpiod dependency which (together with the c++ adapter) is about 60K bytes. It'd be nice to save them, but it means we have to deal with API changes, as it just so happens with V1 and V2 of the gpio character device APIs.

If most chips have only a single gpio chip, I don't know, ours doesn't and just for debugging purpose I try to name all of them :-)
```
gpiochip0  gpiochip1  gpiochip2  gpiochip3  gpiochip4  gpiochip5  gpiochip6
```

Anyhow, is the list above missing anything else for now?

**Reply by dmitrydvorkin** [reply]
gpio chips list are located at
/sys/class/gpio/gpiochip < N >
so Chip or whole module constructor can list them first and then return as the list of options to choose in Node configuration dialog in NodeRed. it will help user to do less typing or make stupid mistake of entering wrong chip name.

I am not good in NodeJS, so giving this idea to you, guys
if there is only one chip (in most cases), you does not have to enter anything and it is already choosen in drop-down node configuration menu. It is usability.

**Comment by dmitrydvorkin** [comment]
> Functionality to return the pinmux configuration (bias, and stuff) and pinmux name
- never cared about this in real projects :)

> Functionality to return the gpio number, e.g. on IMX6 and some others: (chipNumber - 1) * 32 + lineOffset
- same comment

> Functionality to configure Edge detection (with event triggering)

easy:
    struct gpioevent_request ereq; 
    ereq.lineoffset = 0; 
    ereq.eventflags = GPIOEVENT_REQUEST_BOTH_EDGES; 

    ioctl(fd, GPIO_GET_LINEEVENT_IOCTL, &ereq);

three constants GPIOEVENT_REQUEST_* are defined there
https://elixir.bootlin.com/linux/v4.8/source/include/uapi/linux/gpio.h#L112

> Bulk operations
- could be useful, but event triggering gives a lot more.

**Comment by noctarius** [comment]
See, that's how priorities differ. I think event detection is nice, but we're just polling since we basically have modules either being plugged in (and booted) or not. No quick changes.

The issue with event detection is not how to make it happen in the Linux gpio api itself, but those events need to be enqueued into the actual nodejs event loop. And there's nothing like "immediate reaction" to the event since nodejs has pretty hard rules in what order certain types of events are executed.

**Reply by dmitrydvorkin** [reply]
> No quick changes
- button press/release (yes, it is gpio event). May happen rather quick and you can't do simple polling to detect it correctly. Or polling period should be rather small - requires a lot of CPU for that!

GPIO events can be read like this:
    struct gpioevent_data event; 
    read(pin->fd, &event, sizeof(event));

it may be called at any time and it will return the timestamp and event type.
it may be repeated to check if there are more then 1 event is happened - like till "EOF" reading of the file.
You will never loose the butto press/release using this interface and you can call it on "free time" of your app.

I suspect, there are something in NodeRed/NodeJS that can be used as "event queue"

**Reply by noctarius** [reply]
You misunderstood what I meant. Nodejs is single threaded and uses an event loop internally. You can register certain types of events to be executed inside the event loop. But there is a hard order of how the different types of events are executed to prevent dead locks. That means you can generate a trigger event for the button press, but you can't react too it immediately. The button may even not be pressed anymore when event is finally handled (depending on the other code running). For more information see https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/. That means we need to see and think about how the handling in nodejs should happen. Just to make sure it's built in a way to be future proof.

**Reply by dmitrydvorkin** [reply]
yes, thanks for explanation! BTW, with events gpio api we will never loose the event.
so no need very short polling inverval to catch the line switches

**Comment by khill-fbmc** [comment]
I would be happy to help with testing on a physical platform. I [mentioned here](https://github.com/sombriks/node-libgpiod/pull/12#issuecomment-1809103038) that I am working on a RockChip board.

I'd like to chime in and say I would prefer an interface where I can:
```javascript
const pin3 = new Pin(3); // Internally this would map to `new Line(new Chip(), n)`
pin3.requestOutputMode();
// which for me in the instance 
```

It was unclear to me at first how the chip number and line were retrieved, but then found this tool's use:
![image](https://github.com/sombriks/node-libgpiod/assets/107003416/93a32828-1f73-46ca-a1b5-49cf473e94dd)

That was how I found out how to get:
```javascript
const chip = new Chip(1);
const line = new Line(chip, 31);
```

[When I landed on Radxa's doc's for their GPIO](https://docs.radxa.com/en/rock5/rock5a/hardware/rock5a-gpio) I didn't see at first how to get that info, when they were giving the GPIO numbers

**Reply by khill-fbmc** [reply]
By the way, it works as is, I was able to get the first example [here](https://github.com/sombriks/node-libgpiod#status) to blink a LED :)

**Reply by sombriks** [reply]
@khill-fbmc this is a nice proposal, By using pin we would assume the first gpiochip found and map one pin to one line.

the tests will be very appreciate, i am working on how to make github actions proper run tests so we can get more confidence on the current codedabe.

**Reply by khill-fbmc** [reply]
This may not be the best way to do this, but here is how I made a little class to use `gpiofind` and collect all the chip and line numbers

```js
/**
 * Map the gpio pin numbers to their (chip, line) numbers
 */
async init() {
    for (let pinNum of this._pinList) {
        let { output } = await this._gpiofind(pinNum);
        const [chipNum, lineNum] = output.replace("gpiochip", "").split(" ").map(Number);
        this._log(`🔎 Found PIN_${pinNum} on Chip ${chipNum}, Line ${lineNum}`);
        const chip = new Chip(chipNum);
        const line = new Line(chip, lineNum);
        const newPin: GpioPin = {
            pin: {
                num: pinNum,
                formatted: `PIN_${pinNum}`,
                id: `gpiochip${chipNum} ${lineNum}`,
            },
            chip: {
                num: chipNum,
                raw: chip,
            },
            line: {
                num: lineNum,
                raw: line
            },
        };
        this.pins.push(newPin);
    }
}
```

`this._pinList` was assigned in the constructor with a simple `[3,5,7,.... etc]`

This makes getting the `Line` by gpio pin number very easy :)

```js
gpio(pinNum) {
    const pin = this.getPin(pinNum);
    return pin.line.raw;
}

getPin(pinNum: number): GpioPin {
    this._ensureInitComplete();
    const pin = this.pins.find(({ pin }) => pin.num === pinNum);
    if (!pin) throw Error(`PIN_${pinNum} is not a valid pin.`);
    return pin;
}
```

**Reply by khill-fbmc** [reply]
![image](https://github.com/sombriks/node-libgpiod/assets/107003416/3e1b2477-2f1e-4a38-84db-5f9a78ec71c2)

**Comment by noctarius** [comment]
I think the issue is, that GPIO numbers aren't just assigned in a linear fashion to pin numbers. Many SOCs have specific GPIO numbers assigned to specific pads and specific chips.

The most obvious example is a Raspberry Pi, where the pin numbers (on the header) have absolutely nothing to do with the GPIO numbers (see https://pi4j.com/assets/getting-started/pins/headerpins_in_header.png). But this is true for almost all ARM SOCs.

On an i.MX6, as another example, the GPIO number is calculated as 32 + 4 = 36 (meaning port 2/chip1, pin 4/line4). That said, they always assume 32 lines per chip (https://www.digi.com/resources/documentation/digidocs/90001546/reference/bsp/cc6/r_general_purpose_input_output.htm#:~:text=LinuxGPIO_num%20%3D%20(%3Cimx6_gpio_port%3E%20%2D%201)%20*%2032%20%2B%20%3Cimx6_gpio_pin%3E). I'm not sure this would hold true for every chip though.

With the two above examples, the more complicated issue though, what would PIN mean to the user? Is it the header pin? Is it the GPIO number? Is it just a linearly increasing number? If so, how many lines to assume per chip?

**Reply by khill-fbmc** [reply]
For me, I am getting my pin numbers from [here](https://docs.radxa.com/en/rock5/rock5a/hardware/rock5a-gpio), the green ones.

I pass them in as list of pins to "search" for and initialize `Chip`s and `Line`s

```js
export const ROCK5_A_PINS = [
    3,
    5,
    7,
    8,
    10,
    11,
    12,
    13,
    15,
    16,
    18,
    19,
    21,
    22,
    23,
    24,
    26,
    27,
    28,
    29,
    31,
    32,
    33,
    35,
    36,
    38,
    40,
];
```

**Reply by sombriks** [reply]
It is a good idea and very cool design to provide a way to initialize the desired pins upfront, but those pins vary from board to board.

We can discuss ways to do so in a fashion that we don't make the library itself know every single board pin-out.

**Reply by noctarius** [reply]
Right, but that's just the pin header on that specific board. It would mean that everyone would have to set up the mapping. Apart from that how do you map pin 3 to GPIO1_A0? I feel like I'm missing something 🤔

**Reply by khill-fbmc** [reply]
These correlate with the names I saw after running `gpioinfo`

![image](https://github.com/sombriks/node-libgpiod/assets/107003416/3e893852-c194-489e-a680-323c2397288d)

And what my `init` function found

![image](https://github.com/sombriks/node-libgpiod/assets/107003416/8702b7ef-ed4b-4ba4-9c04-9638cc0381e9)

**Reply by noctarius** [reply]
I'd say, that's just because the pins are configured / named in the board's device tree. Other boards won't /may not have them set up.

**Reply by khill-fbmc** [reply]
> [...] that we don't make the library itself know every single board pin-out.

I agree, the library should remain board agnostic

**Reply by khill-fbmc** [reply]
> Right, but that's just the pin header on that specific board. It would mean that everyone would have to set up the mapping. Apart from that how do you map pin 3 to GPIO1_A0? I feel like I'm missing something 🤔

This is all I am doing to get the chip and line numbers, based on my "pin" number

```js
private async _gpiofind(pinNum: number) {
    const { stdout } = await this._exec(`gpiofind PIN_${pinNum}`);
    return {
        output: stdout,
    };
}
```
and

```js
const [chipNum, lineNum] = output.replace("gpiochip", "").split(" ").map(Number);
```

**Reply by noctarius** [reply]
This is where the names are configured. It's unfortunately not a common standard :(
https://github.com/radxa/kernel/blob/d7f6b85f6fcfdf510f382174bf08368e73fb9287/arch/arm64/boot/dts/rockchip/rk3568-rock-3a.dts#L1062

It's not even common to radxa rockchip 3568 boards (which was my initial thought), it's setup specifically set up in the 3A device tree (but I hope it'll match throughout their boards).

**Reply by noctarius** [reply]
Here's the alternative example from a RaspberryPi: https://github.com/radxa/kernel/blob/d7f6b85f6fcfdf510f382174bf08368e73fb9287/arch/arm/boot/dts/bcm2837-rpi-3-b.dts#L61

**Reply by khill-fbmc** [reply]
This was one of my explorations, if it's any consolation? 😄 

```ts
import { execSync } from "node:child_process";
import { Chip, Line } from "node-libgpiod";

export class Pin extends Line {
    constructor(chipNum: number, lineNum: number) {
        super(new Chip(chipNum), lineNum);
    }

    static gpio(pinNum: number): Pin {
        const stdout = execSync(`gpiofind PIN_${pinNum}`);
        const cleaned = String(stdout).replace("gpiochip", "");
        const [chipNum, lineNum] = cleaned.split(" ").map(Number);
        console.log(`🔎 Found PIN_${pinNum} on Chip ${chipNum}, Line ${lineNum}`);
        return new Pin(chipNum, lineNum);
    }
}
```

and then this worked to blink an LED

```ts
const gpio3 = Pin.gpio(3);
gpio3.requestOutputMode();
gpio3.setValue(0);

let count = 10;
const blink = () => {
    if (count) {
        gpio3.setValue(count-- % 2);
        setTimeout(blink, 1000);
    } // else line.release();
    // not needed, libgpiod releases resources on process exit
};

setTimeout(blink, 1000);
```

**Reply by khill-fbmc** [reply]
> Here's the alternative example from a RaspberryPi: https://github.com/radxa/kernel/blob/d7f6b85f6fcfdf510f382174bf08368e73fb9287/arch/arm/boot/dts/bcm2837-rpi-3-b.dts#L61

I don't know that file format, but I can see `GPIO23` instead of things like `PIN_3` so that makes sense that on a pi you couldn't use my solution. I assume on a pi you would have to `gpiofind GPIO23`?

**Reply by noctarius** [reply]
Yes, for example. But it also means your split magic wouldn't just work. Unfortunately, I don't think there's any way to do it. Many boards (especially custom ones) will not have any names set up, because "why?".

I like the simplified version, but I wouldn't know how to make it work in a generic way, and adding something which works on one board, or if you pass in the name config (which doesn't make things easier and you wouldn't even have to, you can build the table from the Linux kernel lookups), I'd still only works on board with the gpio lines named.

PS: The file format is the device tree descriptor. That file describes the board and how components are interconnected. Since there isn't a BIOS or something that can discover components (like in a PC) and you normally have hardwired connections (so discovering every boot wouldn't make sense), Linux uses device trees to learn about the components. Every ARM board uses them, (I think) every RISC-V board uses them, too. Here's some info: https://docs.kernel.org/devicetree/usage-model.html

**Reply by khill-fbmc** [reply]
Thank you for such a detailed reply. It makes sense.

I had the idea to make a wrapper around your library to extend it with board specific mappings, so maybe I will think more on that.

**Reply by noctarius** [reply]
My pleasure and no worries, before I had my own startup working directly with ARM boards and Linux (DTS), I had neither knew. The only thing I knew was that the raspi has those "weird" overlays to enable hardware extensions. Now I know why. They add additional components to the device tree :-)

**Reply by khill-fbmc** [reply]
I recently came into possession of a Pico Pi H, so I am learning MicroPython alongside the Rock5 as well :)

**Comment by d-beamon** [comment]
Hi!
It would be great to support all the line configuration flags that libgpiod supports. Mainly when configuring a line as an input line.
For instance:
- gpiod_line_edge
- gpiod_line_bias
- gpiod_line_drive

**Comment by sombriks** [comment]
long time i don't comment, but this is in progress!

my last stop was on callbacxk monitoring, which would be a nice addition to 1.x version before move on to 2.x
